Src/VimCore/KeyMap.fs

```diff
@@ -2,11 +2,47 @@
 
 namespace Vim
 open System
+open System.Collections.Generic
+open System.Diagnostics
 open Vim.Interpreter
+open CollectionExtensions
+open Microsoft.VisualStudio.Text
 
-/// Map of a LHS of a mapping to the RHS.  The bool is used to indicate whether or not 
-/// the RHS should be remapped as part of an expansion
-type RemapModeMap = Map<KeyInputSet, KeyMapping>
+type internal MappingMap<'TMode, 'TData when 'TMode : equality>(_modeList: 'TMode list) =
+
+    let _map = Dictionary<'TMode, Dictionary<KeyInputSet, 'TData>>(capacity = _modeList.Length)
+
+    do
+        for mode in _modeList do
+            _map.[mode] <- Dictionary<KeyInputSet, 'TData>()
+
+    member x.Mappings: 'TData seq = 
+        seq {
+            for kvp in _map.Values do
+                yield! kvp.Values
+        }
+
+    member x.Add(key, mode, data) =
+        let map = _map.[mode]
+        map.[key] <- data
+
+    member x.Remove(key, mode) =
+        let map = _map.[mode] 
+        map.Remove key
+
+    member x.Get(key, mode) =
+        let map = _map.[mode]
+        map.TryGetValueEx key
+
+    member x.Get(mode) = _map.[mode].Values :> 'TData seq

```

Comment:
The two advantages I see to tuple based arguments are overloading and named arguments at the call site. The latter is more interesting to me. 

This started coming up during the code point PR, IIRC, where I found I wanted to be able to better distinguish between columns and spaces at call sites. This naturally fell out for constructors because they are forced to be tupled. But for factory methods and normal methods I couldn't be as precise. Since then I've increasingly wanted to use named arguments. 

Recently reading through the F# documentation I found that the recommended way is tuple for `member` and curried for `let`. That has pushed me over the edge to a degree. 

---
In reply to: [306000157](https://github.com/VsVim/VsVim/pull/2689#discussion_r306000157) [](ancestors = 306000157)